gold-digger/src/firebaseModel.js

```diff
@@ -1,25 +1,148 @@
+    // TODO implement
+    // GET from firebase
 }
 
-function persistenceToModel(persistedData, model) {
-    // TO-DO
+
+function firebaseModelPromise(model, setModel) {
+    let userPATH="";
+    if(model.userid) {
+        userPATH = "/"+model.userid;
+        console.log("FIREBASE userpath set to: " + userPATH);
+    } else {
+        console.log("FIREBASE No user set in model, meaning firebasepath is incorrect");
+    }
+    
+    // Retrieves persisted model parameters
+    
+    return get(ref(db, PATH+userPATH+"-modelParams")).then(toModelACB).then(addObserversACB);
+
+    // Saves any persisted data into the model (received as parameter)
+    function toModelACB(dataFromFirebase) {
+        return persistenceToModelParams(dataFromFirebase.val(), model, setModel);
+    }
+    
+    function addObserversACB() {
+        console.log("observers added");
+        model.addObserver(obsGeneralParamsACB);
+        model.addObserver(obsGeneratedListACB);
+        model.addObserver(logOutACB);
+
+        // TODO check if any of these are needed (it doesn't seem like it at the moment)
+        //setModel(model);
+        //return model;
+    }
+
+    
+    // Observes all model parameters and saves to firebase
+    function obsGeneralParamsACB(payload){
+        console.log("obsGeneralParamsACB notified!");
+        if (payload.key) {
+            if(payload.key === "modelParams") {
+                set(ref(db, PATH+userPATH+"-modelParams"), modelParamsToPersistence(model));
+            }
+        }
+    }
+
+    // Observes current generated list and saves any changes to firebase
+    // (eg changed name, removed song)
+    function obsGeneratedListACB(payload){
+        if (payload.key) {
+            if(payload.key === "editGenerated") {
+                /*
+                TODO
+                - get key (aka firebasespecific ID) of current generated list
+                - something like payload.firebaseKey, alt model.generated.firebasekey
+                */
+                const firebaseKey = 0;
+                set(ref(db, PATH+userPATH+"-"+firebaseKey), generatedListToPersistence(model));
+            }
+        }
+    }
+
+
+    function logOutACB (payload) {
+        if (payload.key) {
+            if (payload.key === "logout") {
+                logout(model);
+            }
+        }
+    }
+
+
+    function logout(model) {
+        // remove all observers before resetting model
+        model.removeObserver(obsGeneralParamsACB);
+        model.removeObserver(obsGeneratedListACB);
+        model.removeObserver(logOutACB);
+
+        // TODO reset all values in model
+        model.user = null;
+
+        console.log("Observers removed and model reset");
+    }
 }
 
-function firebaseModelPromise(model) {
-    // TO-DO
-}
\ No newline at end of file
+
+export {modelParamsToPersistence, generatedListToPersistence, persistenceToModelParams, persistenceToGeneratedList, generatedListPromise, firebaseModelPromise};
```

Comment:
They're not used otherwhere right?

Reply:
Some of them are, some of them will be, some of them won't be. I'll keep an eye on them